1D non-directed rotor scan troubleshooting

xiaoruiDong opened the pull request:
This is a complete update and rewrite of #263. 
This PR aims to provide customized solutions to the 1D non-directed rotor scan. Currently, it only supports Gaussian scan results, but if any other ESS provides internal coordinate (IC) Infos, it is not difficult to implement this strategy. 

The methodology applied is based on the identification of abnormal behavior of internal coordinates (bond, angle, dihedral). Pretty intuitively, if any of the ICs have a sudden significant change or a difference between the initial and final conformer, these ICs tend to be relaxed too much, resulting in switching to a different conformer (not accounting the change of the scanning dihedral) or even different species (bond breaking) during the scan. This update is able to identify these changes by implementing thresholds (check If any of the changes exceed the threshold). Usually, the results from identification are redundant (you don't wish that many dihedrals!), so an algorithm is also implemented to yield a clean solution. This is important, especially when dealing with bond flipping dihedrals which share the same pivots as the scanning dihedrals.

This PR includes changes in 4 sections:
1. parser.py: Add parsers to allow parsing IC infos
2. common.py: Add functions to allow quick IC difference check and same pivot check
3. job/trsh.py: Update the methods.
4. scheduler.py: Changes based on the change of trsh.py

Unit tests are added for some of the modules. It is also tested on a 2k scan results set. Primarily, the method can successfully come up with solutions for each problematic job identified and the solutions make sense. But tests haven't been done to see if all the solution perfectly solve the problem. Even not, the new method now makes multiple-time troubleshooting possible and should solve the problems eventually.

alongd commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 16
           

Clones added
============
- arc/job/trsh.py  4
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/ReactionMechanismGenerator/ARC/pullRequest?prid=5069484&bid=16680380)

xiaoruiDong commented:
Unused import will be removed. I checked the travis test and found a problem related to the sequence of the list (obtain [b,a], expect[a,b], not equal). The error did not show up in my local test. Any advice?

alongd commented:
One of the tests passes, and another failed (and also, locally it did pass for you).
Looks like something in the code might be nondeterministic, let's take a look.

alongd commented on arc/common.py:
```diff
@@ -907,3 +907,43 @@ def time_lapse(t0):
     return f'{d}{h:02.0f}:{m:02.0f}:{s:02.0f}'
+
+
+def is_same_pivot(dihed1, dihed2):
```
let's try to be consistent with the rest of ARC in terms of naming.
"dihedral" sounds like the value of the angle, let's use "torsion" for describing the atom indices defining the dihedral angle. (see, e.g., https://github.com/ReactionMechanismGenerator/ARC/blob/26d7daed170880a58669255512f66370109de173/arc/species/vectors.py#L177)
I use "scan" as a parameter name for 1-indexed torsions (like the input to Gaussian), and "torsion" as a parameter name for 0-indexed torsions. It helps to know how to work with each parameter just by its name

xiaoruiDong replied:
@alongd Help! There is something that needs to be clarified. To me, [1, 2, 3, 4] and [5, 2, 3, 6] refers to the same torsion. But in scan troubleshooting, it is important to distinguish them, since they are not changed with the same degrees in some cases. Is that okay to call them torsion1 and torsion2? To me, they just refer to different dihedral angles, that's the reason I call them dihedrals.

alongd replied:
Yes, it's OK to call them torsion1 and torsion2, even if they (may) represent the same torsion, since technically the answer the above defiinition.

alongd commented on arc/common.py:
```diff
@@ -907,3 +907,43 @@ def time_lapse(t0):
+        dihed2 (list): The four atom indices representing the second torsion.
+
+    Returns:
+        (Bool): True if two dihedrals share the same pivot.
```
I think there's no need for `( )` around `bool`, also I think it should be lowercase
Add "``True``" (please see the source for this comment, these are two weird quotation marks on each side) around `True`, it shows it in _italics_ in the API documentation this way.

alongd commented on arc/common.py:
```diff
@@ -907,3 +907,43 @@ def time_lapse(t0):
+    """
+    if len(parent_list) < len(child_list):
+        return False
+    for item in child_list:
```
consider
`if any([item not in parent_list for item in child_list])`

alongd commented on arc/parser.py:
```diff
@@ -10,11 +10,12 @@
 import qcelemental as qcel
 import numpy as np
+import pandas as pd
 from arkane.exceptions import LogError
```
Add an empty line between general imports and Arkane/RMG

alongd commented on arc/parser.py:
```diff
@@ -10,11 +10,12 @@
 from arkane.util import determine_qm_software
 
-from arc.common import get_logger, determine_ess
+from arc.common import get_logger, determine_ess, is_same_pivot
```
do you mind sorting these alphabetically?

alongd commented:
Adding a screenshot of the error in case it passes when force-pushed:
![image](https://user-images.githubusercontent.com/16158262/75411605-d0b15f00-58ed-11ea-9aa8-2677e9853336.png)

xiaoruiDong commented on arc/parser.py:
```diff
@@ -718,3 +718,58 @@ def process_conformers_file(conformers_path):
+                        tail_repeat = 0
+                        mode = 'read'
+                        blks.append([])
+                        blks[-1].append(line)
```
blks is a list of lists of lines. E.g., `blks = [['good', 'morning',], ['good', 'bye'],]`. So `blks[-1].append` makes sure that line is apppend to the last block (E.g.,`['good', 'bye']` instead of `['good', 'morning',]`)

xiaoruiDong commented on arc/parser.py:
```diff
@@ -718,3 +718,58 @@ def process_conformers_file(conformers_path):
+            line = f.readline()
+        # Remove the last incomplete search
+        if len(blks) > 0 and (tail_repeat != tail_count):
+            blks.pop()
```
Since `blks` is a list of blocks, this pops the last block.

xiaoruiDong commented on arc/parser.py:
```diff
@@ -773,3 +773,47 @@ def search(x, y): return x in y
         return blks
+
+
+def parse_scan_args(file_path, software=None):
```
My suggestion is keeping them separate, considering maintainance and flexibility.

xiaoruiDong commented on arc/parser.py:
```diff
@@ -773,3 +773,47 @@ def search(x, y): return x in y
+
+def parse_scan_args(file_path, software=None):
+    """
+    Get the scan arguments, including which internal coordinates are scanning, which are frozen, 
```
You are right

xiaoruiDong commented on arc/parser.py:
```diff
@@ -773,3 +773,47 @@ def search(x, y): return x in y
+                 'step': 0, 'step_size': 0, 'n_atom': 0}
+    if software == 'gaussian':
+        scan_blk = parse_str_blocks(file_path, 'The following ModRedundant input section has been read:',
+                                    'Isotopes and Nuclear Properties', regex=False)[0][1:-1]
```
`[1:-1]` exclude the first and the last lines, while `[1:]` only excludes the first line.

alongd commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 1
           

Clones added
============
- arc/job/trsh.py  4
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/ReactionMechanismGenerator/ARC/pullRequest?prid=5069484&bid=16680380)

xiaoruiDong commented:
Thank you @alongd for the preliminary review. I know you are busy, so take your time to review the rest of the code. I think the indeterministic behavior is due to the use of `list(set())` (to remove duplicates) which loses sequence information, but the sequence has no importance for post-processing. So I just modify the way of testing.

alongd commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 1
           

Clones added
============
- arc/job/trsh.py  4
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/ReactionMechanismGenerator/ARC/pullRequest?prid=5069484&bid=16680380)

alongd commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 1
           

Clones added
============
- arc/job/trsh.py  4
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/ReactionMechanismGenerator/ARC/pullRequest?prid=5069484&bid=16680380)

alongd commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 1
           

Clones added
============
- arc/job/trsh.py  4
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/ReactionMechanismGenerator/ARC/pullRequest?prid=5069484&bid=16680380)

alongd commented on arc/job/trsh.py:
```diff
@@ -978,57 +1097,184 @@ def scan_quality_check(label, pivots, energies, scan_res=rotor_scan_resolution,
+                return invalidate, invalidation_reason, message, actions
+        else:
+            bond_break = [scan_conformers['atoms'][ic_label] for ic_label in list(set(bond_break))]
+            invalidation_reason = f'Bond {bond_break} break during the scan.'
```
"broke" instead of "break"

alongd commented on arc/job/trsh.py:
```diff
@@ -978,57 +1097,184 @@ def scan_quality_check(label, pivots, energies, scan_res=rotor_scan_resolution,
+        else:
+            bond_break = [scan_conformers['atoms'][ic_label] for ic_label in list(set(bond_break))]
+            invalidation_reason = f'Bond {bond_break} break during the scan.'
+            message = f'Rotor scan of {label} between pivots {pivots} has breaking bonds ' \
```
"broken" instead of "breaking"

alongd commented on arc/job/trsh.py:
```diff
@@ -978,57 +1097,184 @@ def scan_quality_check(label, pivots, energies, scan_res=rotor_scan_resolution,
+            invalidation_reason = 'Inconsistent initial and final conformers'
+            message = f'Rotor scan of {label} between pivots {pivots} has inconsistent initial ' \
+                      f'and final conformers. Internal coordinates {changed_ic_dict[0]} are different.' \
+                      f'Invalidating rotor.'
```
A general comment: I think it's incorrect to say "invalidating rotor" if we plan to troubleshoot it with some actions. "invalidation" is for rotors we don't know how to treat (if all trsh methods failed), and adding them to the statmech analysis will do more harm than good. Perhaps say "ARC will attempt to troubleshoot this rotor." instead.

alongd commented on arc/job/trsh.py:
```diff
@@ -978,57 +1097,184 @@ def scan_quality_check(label, pivots, energies, scan_res=rotor_scan_resolution,
+                      f'between two consecutive conformers. Invalidating rotor.'
+            changed_ic_label = []
+            for index, ics in changed_ic_dict.items():
+                if index > 0: # Do not include the initial/final differences which may include more ics
```
minor: two spaces between code and `#` comment marker

alongd commented on arc/job/trsh.py:
```diff
@@ -978,57 +1097,184 @@ def scan_quality_check(label, pivots, energies, scan_res=rotor_scan_resolution,
+                    actions = ['inc_res', 'all_freeze']
+                return invalidate, invalidation_reason, message, actions
+
+        # 2.3 Check conformation:
```
"# 2.3 Check energy and change conformation if needed"

alongd commented on arc/job/trshTest.py:
```diff
@@ -448,6 +450,82 @@ def test_trsh_negative_freq(self):
+        # Case 1: non-smooth scan which troubleshot once
+        case1 = {'label': 'CH2OOH',
+                 'pivots': [1, 2],
+                 'energies': np.array([1.05018884e-05, 5.19029816e-01, 1.98559728e+00, 4.14500528e+00,
```
do we need to specify the energies if we have the log file? Perhaps parse them instead? (same below)

alongd commented on arc/job/trshTest.py:
```diff
@@ -448,6 +450,82 @@ def test_trsh_negative_freq(self):
+                                    1.09528319e+01, 1.20173539e+01, 1.21102100e+01, 1.11961343e+01,
+                                    9.36754720e+00, 6.92812177e+00, 4.32620415e+00, 2.05935807e+00,
+                                    5.33139252e-01, 0.00000000e+00]),
+                 'scan_res': 4.0,
```
there are 46 points, seems like a resolution of 8 degrees

alongd commented on arc/job/trshTest.py:
```diff
@@ -448,6 +450,82 @@ def test_trsh_negative_freq(self):
+        expect_message = 'Rotor scan of CH2OOH between pivots [1, 2] is inconsistent between ' \
+                         'two consecutive conformers. Invalidating rotor. Between conformer ' \
+                         '#63 and #64, D3,D2 are changed significantly. Between conformer #80 ' \
+                         'and #81, D3,D2 are changed significantly.'
```
how come there are more than 80 conformers if there are only 46 energy points?

alongd commented on arc/parser.py:
```diff
@@ -718,3 +720,241 @@ def process_conformers_file(conformers_path):
     return xyzs, energies
+
+
+def parse_str_blocks(file_path, head_pat, tail_pat, regex=True, tail_count=1, block_count=1):
```
Could you add type hints throughout this PR in the following format? (If for some functions it's not trivial, feel free to skip it!)
(for regex see https://docs.python.org/3/library/typing.html#typing.Match)
E.g.:
```python
from typing import Match

def parse_str_blocks(file_path: str,
                     head_pat: str or Match,
                     tail_pat: str or Match,
                     regex: bool = True,
                     tail_count: int = 1,
                     block_count: int = 1,
                     ) -> list:
```

alongd commented on arc/parser.py:
```diff
@@ -718,3 +720,241 @@ def process_conformers_file(conformers_path):
+        for line in ic_info_block:
+            terms = line.split()
+            ic_dict['label'].append(terms[1])
+            ic_dict['type'].append(terms[1][0])  #'R: bond, A: angle, D: dihedral
```
minor: space after `#`

alongd commented on arc/parser.py:
```diff
@@ -718,3 +720,241 @@ def process_conformers_file(conformers_path):
+
+            # Identify redundant, cases like 5 atom angles or redundant atoms
+            if (ic_dict['type'][-1] == 'A' and len(atom_inds) > 3) \
+            and (ic_dict['type'][-1] == 'R' and len(atom_inds) > 2) \
```
fix indentation (two more "tabs")

alongd commented on arc/parser.py:
```diff
@@ -718,3 +720,241 @@ def process_conformers_file(conformers_path):
+            # Identify redundant, cases like 5 atom angles or redundant atoms
+            if (ic_dict['type'][-1] == 'A' and len(atom_inds) > 3) \
+            and (ic_dict['type'][-1] == 'R' and len(atom_inds) > 2) \
+            and (ic_dict['type'][-1] == 'D' and len(atom_inds) > 4):
```
I think these `and`s should be `or`s
```python
if (ic_dict['type'][-1] == 'A' and len(atom_inds) > 3) \
        or (ic_dict['type'][-1] == 'R' and len(atom_inds) > 2) \
        or (ic_dict['type'][-1] == 'D' and len(atom_inds) > 4):
```

alongd commented on arc/parser.py:
```diff
@@ -718,3 +720,241 @@ def process_conformers_file(conformers_path):
+                ic_dict['redundant'].append(True)
+            else:
+                weird_indices = [index for index in ic_dict['atoms']
+                                [-1] if index <= 0 or index > max_atom_ind]
```
does `[-1]` refer to `ic_dict['atoms']`? Try not to break them for better readability

alongd commented on arc/parser.py:
```diff
@@ -718,3 +720,241 @@ def process_conformers_file(conformers_path):
+                ic_dict['redundant'].append(True)
+            else:
+                weird_indices = [index for index in ic_dict['atoms']
+                                [-1] if index <= 0 or index > max_atom_ind]
```
what's the meaning of such indices in a Gaussian output file?

xiaoruiDong replied:
The role of these redundant atoms is not clear (to me). However, they are found in the internal coordinates. The weirdest case I found is that Gaussian used 5 atoms to define an angle that I even cannot figure out which angle it is.

alongd commented on arc/parser.py:
```diff
@@ -718,3 +720,241 @@ def process_conformers_file(conformers_path):
+    ic_dict = {item: [] for item in ['label', 'value']}
+    if software == 'gaussian':
+        for line in ic_block:
+            terms = line.split()
```
could you give a commented example of the format of the `line` this code in processing?

alongd commented on arc/parserTest.py:
```diff
@@ -332,6 +332,34 @@ def test_process_conformers_file(self):
         spc6 = ARCSpecies(label='tst6', xyz=path3)
         self.assertEqual(len(spc6.conformers), 4)
 
+    def test_parse_str_blocks(self):
```
Can you also add tests for parse_scan_conformers(), parse_ic_values(), and parse_ic_info()?

alongd commented on arc/scheduler.py:
```diff
@@ -2187,16 +2187,27 @@ def check_scan_job(self, label, job):
-                        self.species_dict[label].set_dihedral(
-                            scan=self.species_dict[label].rotors_dict[i]['scan'],
-                            deg_increment=deg_increment)
+                        if len(actions) == 3:
```
I think we should make `actions` a dictionary, otherwise this is a recipe for bugs. Do you mind converting it?

alongd commented on arc/species/species.py:
```diff
@@ -1063,36 +1063,38 @@ def set_dihedral(self, scan, deg_increment=None, deg_abs=None, count=True, xyz=N
-        if deg_increment is not None and not isinstance(deg_increment, float):
+        if deg_increment is None and deg_abs is None:
+            raise InputError('Either deg_increment or deg_abs must be specified.')
+        elif deg_increment is not None and not isinstance(deg_increment, float):
```
Please modify to `isinstance(deg_increment, (float, int))` (below as well, we can keep the error message with the `float` requirement.

alongd replied:
not need for an `elif` after a `raise`

alongd commented:
Take a look at the Codacy report, it complains about an unused import

alongd commented:
I ran `COCOOC` on this branch, it seems to be oscillating when troubleshooting two of the rotors:
```
Found 5 possible rotors for COCOOC
Pivot list(s) for COCOOC: [[1, 2], [2, 3], [3, 4], [4, 5], [5, 6]]

...


Trying to troubleshoot rotor [3, 4] of COCOOC using ['freeze', [[3, 4, 5, 6]]]...
Running job scan_a40277 for COCOOC (pivots: [3, 4])
  Ending job scan_a40276 for COCOOC (run time: 0:12:13)
Error: Rotor scan of COCOOC between pivots [2, 3] is inconsistent between two consecutive conformers. Invalidating rotor. Between conformer #25 and #26, D10 are changed significantly. Between conformer #26 and #27, D10 are changed significantly. Between conformer #32 and #33, D10 are changed significantly.
Trying to troubleshoot rotor [2, 3] of COCOOC using ['freeze', [[3, 4, 5, 6]]]...
Running job scan_a40278 for COCOOC (pivots: [2, 3])
  Ending job scan_a40277 for COCOOC (run time: 0:13:07)
Error: Rotor scan of COCOOC between pivots [3, 4] is inconsistent between two consecutive conformers. Invalidating rotor. Between conformer #34 and #35, D10 are changed significantly. Between conformer #41 and #42, D10 are changed significantly.
Trying to troubleshoot rotor [3, 4] of COCOOC using ['freeze', [[3, 4, 5, 6]]]...
Running job scan_a40279 for COCOOC (pivots: [3, 4])
  Ending job scan_a40278 for COCOOC (run time: 0:12:18)
Error: Rotor scan of COCOOC between pivots [2, 3] is inconsistent between two consecutive conformers. Invalidating rotor. Between conformer #25 and #26, D10 are changed significantly. Between conformer #26 and #27, D10 are changed significantly. Between conformer #32 and #33, D10 are changed significantly.
Trying to troubleshoot rotor [2, 3] of COCOOC using ['freeze', [[3, 4, 5, 6]]]...
Running job scan_a40280 for COCOOC (pivots: [2, 3])
  Ending job scan_a40279 for COCOOC (run time: 0:13:17)
Error: Rotor scan of COCOOC between pivots [3, 4] is inconsistent between two consecutive conformers. Invalidating rotor. Between conformer #34 and #35, D10 are changed significantly. Between conformer #41 and #42, D10 are changed significantly.
Trying to troubleshoot rotor [3, 4] of COCOOC using ['freeze', [[3, 4, 5, 6]]]...
Running job scan_a40281 for COCOOC (pivots: [3, 4])
```

xiaoruiDong commented:
> I ran `COCOOC` on this branch, it seems to be oscillating when troubleshooting two of the rotors:
> 
> ```
> Found 5 possible rotors for COCOOC
> Pivot list(s) for COCOOC: [[1, 2], [2, 3], [3, 4], [4, 5], [5, 6]]
> 
> ...
> 
> 
> Trying to troubleshoot rotor [3, 4] of COCOOC using ['freeze', [[3, 4, 5, 6]]]...
> Running job scan_a40277 for COCOOC (pivots: [3, 4])
>   Ending job scan_a40276 for COCOOC (run time: 0:12:13)
> Error: Rotor scan of COCOOC between pivots [2, 3] is inconsistent between two consecutive conformers. Invalidating rotor. Between conformer #25 and #26, D10 are changed significantly. Between conformer #26 and #27, D10 are changed significantly. Between conformer #32 and #33, D10 are changed significantly.
> Trying to troubleshoot rotor [2, 3] of COCOOC using ['freeze', [[3, 4, 5, 6]]]...
> Running job scan_a40278 for COCOOC (pivots: [2, 3])
>   Ending job scan_a40277 for COCOOC (run time: 0:13:07)
> Error: Rotor scan of COCOOC between pivots [3, 4] is inconsistent between two consecutive conformers. Invalidating rotor. Between conformer #34 and #35, D10 are changed significantly. Between conformer #41 and #42, D10 are changed significantly.
> Trying to troubleshoot rotor [3, 4] of COCOOC using ['freeze', [[3, 4, 5, 6]]]...
> Running job scan_a40279 for COCOOC (pivots: [3, 4])
>   Ending job scan_a40278 for COCOOC (run time: 0:12:18)
> Error: Rotor scan of COCOOC between pivots [2, 3] is inconsistent between two consecutive conformers. Invalidating rotor. Between conformer #25 and #26, D10 are changed significantly. Between conformer #26 and #27, D10 are changed significantly. Between conformer #32 and #33, D10 are changed significantly.
> Trying to troubleshoot rotor [2, 3] of COCOOC using ['freeze', [[3, 4, 5, 6]]]...
> Running job scan_a40280 for COCOOC (pivots: [2, 3])
>   Ending job scan_a40279 for COCOOC (run time: 0:13:17)
> Error: Rotor scan of COCOOC between pivots [3, 4] is inconsistent between two consecutive conformers. Invalidating rotor. Between conformer #34 and #35, D10 are changed significantly. Between conformer #41 and #42, D10 are changed significantly.
> Trying to troubleshoot rotor [3, 4] of COCOOC using ['freeze', [[3, 4, 5, 6]]]...
> Running job scan_a40281 for COCOOC (pivots: [3, 4])
> ```

Thank you for the feedback. To me, this seems to be that the frozen is not actually done. If possible, can you please screenshot the scan info of the input file of scan_a40279 and scan_a40278. D(3, 4, 5, 6) is expected to be frozen and no longer show up.

alongd commented:
Sure, here are the input files:
[scan_a40278.txt](https://github.com/ReactionMechanismGenerator/ARC/files/4325322/scan_a40278.txt)
[scan_a40279.txt](https://github.com/ReactionMechanismGenerator/ARC/files/4325323/scan_a40279.txt)

xiaoruiDong commented on arc/parser.py:
```diff
@@ -718,3 +718,58 @@ def process_conformers_file(conformers_path):
+                        mode = 'search'
+            line = f.readline()
+        # Remove the last incomplete search
+        if len(blks) > 0 and (tail_repeat != tail_count):
```
For example, you only find the head, but no tail is found until the last line. Then, it may be desirable to remove this block, since it does not satisfy the search pattern.

xiaoruiDong commented on arc/common.py:
```diff
@@ -947,3 +947,75 @@ def time_lapse(t0):
+    # First iteration without 180/-180 adjustment
+    d_change = (diheds[ind2] - diheds[ind1] - delta).abs() > threshold
+    # Apply 180/-180 adjustment to those shown significance
+    for d_label in d_change[d_change == True].index:
```
When there is only one column (with index) in the DataFrame., you can use this syntax as a filter function. It will return all of the entries which equal to `True`. So in this context, this will return all torsions which have significant differences. Look at this link about pandas syntax for filtering, http://scrapingauthority.com/pandas-dataframe-filtering/

xiaoruiDong commented on arc/job/trsh.py:
```diff
@@ -468,31 +469,143 @@ def trsh_scan_job(label, scan_res, scan, species_scan_lists, methods):
-            raise TrshError(f'Could not find the dihedral to troubleshoot for in the scan list of species {label}')
+        try:
+            scan_args = parse_scan_args(log_file)
+            already_frozen = scan_args['freeze']
```
`scan_args` always has `freeze` but yes, you are right, it should be outside.

xiaoruiDong commented on arc/job/trsh.py:
```diff
@@ -468,31 +469,143 @@ def trsh_scan_job(label, scan_res, scan, species_scan_lists, methods):
+            # Find the one shares three common atoms with the scanning one
+            for ic in methods[1]:
+                if ic[1:] == scan[1:]:
+                    # scan = [1, 2, 3, 4],  ic = [5, 2, 3, 4], freeze[5, 1, 2, 3] to avoid flip
```
In a real scan, `[1, 2, 3, 4]` does not change equally as `[5, 2, 3, 4]`. In most cases, the changes are quite close. However, if one of the pivot atom is a radical site, the difference can be quite huge yielding non-smoothness in the PES. However, we cannot freeze say `[5, 2, 3, 4]`, since we are rotating `[1, 2, 3, 4]`, otherwise `5` and `1` will overlap at some point and fail the scan job. So, I have to come up with a special treatment.

xiaoruiDong commented on arc/job/trsh.py:
```diff
@@ -468,31 +469,143 @@ def trsh_scan_job(label, scan_res, scan, species_scan_lists, methods):
+                    break
+                elif is_same_pivot(ic, scan):
+                    same_pivot.append(ic)
+            else:
```
This `else` is for cases where no three-atom-identical torsion is found.  For this case, freeze both tops.

xiaoruiDong commented on arc/job/trsh.py:
```diff
@@ -978,57 +1097,184 @@ def scan_quality_check(label, pivots, energies, scan_res=rotor_scan_resolution,
+        bond_break = []
+        # Exclude those with boken bonds (different species)
+        for ics in changed_ic_dict.values():
+            any_bond_break = [ic for ic in ics if 'R' in ic]
```
Identified problematic internal coordinates are labeled as `R(X,Y)`, `A(X,Y,Z)` and `D(W,X,Y,Z)`. For all the cases I've ever seen, `R(X,Y)` always refers to a bond in a stable molecule (though TS may be tricky). So, if any bond 'R(X,Y)' are found changing significantly, the bond is almost certain to be broken.

xiaoruiDong commented:
@alongd, I finally corrected all the things you mentioned. Thank you for being so patient to look at my codes. I tried to make the followups clean. I guess most fixup commits are fine and clear, but the ones for trsh.py and schedular.py might be messy. I tried to add as many comments as I can, so for those who are not familiar with this part can understand the workflow. I also planned to add some comments to help you review. BTW, I also cherry-picked one of your commit in another branch, if it is not okay, please let me know. Regarding the test you have done, that is simply because the line of 'scan' has an extra `'\n'` since about 3 months ago, and I am not aware of that. By removing the starting `'\n'` in the `scan_trsh`, I think everything should work perfectly fine.

xiaoruiDong commented on arc/parser.py:
```diff
@@ -265,6 +265,40 @@ def parse_1d_scan_energies(path):
     return energies, angles
 
 
+def parse_1d_scan_coords(path):
```
I  cherry-picked your commits so that I can directly get the conformer xyz instead of using set_dihedrals() to change the current xyz.

alongd replied:
no problem

xiaoruiDong commented on arc/job/trsh.py:
```diff
@@ -626,6 +615,69 @@ def trsh_scan_job(label: str,
+                       ) -> list:
+    """
+    Troubleshoot special rotor cases given all problematic torsional internal
+    coordinates. Special rotors include rotor to be scanned and rotor already frozen.
```
If this is not clear, this method is not to troubleshoot 'the rotor to be scanned' or the 'rotor already frozen, it is to troubleshoot rotors which are relevant to these special rotors. Example: `scan = [1, 2, 3, 4]`, however, `scan_quality_check()` finds `[1, 2, 3, 5]` is problematic. We cannot freeze `[1, 2, 3, 5]` which will for sure lead to issues. By using this function, it will figure out which internal coordinates we need to actually freeze to solve the problem.

alongd replied:
I see. Can you give this example in the docstring?

alongd replied:
was this comment addressed?

xiaoruiDong replied:
Examples are included in the docstring. Thanks.

xiaoruiDong commented on arc/job/trsh.py:
```diff
@@ -25,14 +25,16 @@
                           inconsistency_az,
                           list_available_nodes_command,
                           maximum_barrier,
+                          preserve_param_in_scan_stable,
```
Named something similar to what you used in another branch

xiaoruiDong commented on arc/job/trsh.py:
```diff
@@ -1147,23 +1149,22 @@ def scan_quality_check(label: str,
     message, invalidation_reason = '', ''
     invalidate = False
-    actions = list()
+    actions = dict()
```
Changed to `dict` as you suggested

xiaoruiDong commented on arc/job/trsh.py:
```diff
@@ -1176,122 +1177,117 @@ def scan_quality_check(label: str,
             conformer_based_check = True
 
-    if conformer_based_check:
-        # 1. Check based on intermediate conformers
```
Moved to `settings.py`

xiaoruiDong commented on arc/job/trsh.py:
```diff
@@ -1176,122 +1177,117 @@ def scan_quality_check(label: str,
-        # Can be moved to settings.py later
-        th_bond = 0.1
-        th_angle = 10
-        th_dihed = 20
```
Moved `settings.py`

xiaoruiDong commented on arc/job/trsh.py:
```diff
@@ -1176,122 +1177,117 @@ def scan_quality_check(label: str,
-                                    if is_same_pivot(deg_abs[-1][0], scan_conformers['atoms'][d_l])]
-                actions = ['change conformer', deg_abs]
+                conf_xyzs = parse_1d_scan_coords(log_file)
+                actions = {'change conformer': conf_xyzs[min_index]}
```
New behavior: using the xyz of the conformer.

xiaoruiDong commented on arc/job/trsh.py:
```diff
@@ -1312,18 +1309,30 @@ def scan_quality_check(label: str,
-                    actions = ['inc_res']
-                elif ['inc_res', 'all_freeze'] not in used_methods:
-                    actions = ['inc_res', 'all_freeze']
+                # Propose a method
```
`Used methods` are now changed to a list of `dict`s. The checking is a little bit complicated. We can simplify it to `actions = {'inc_res': None, 'freeze': 'all'}` in the future.

xiaoruiDong commented on arc/job/trsh.py:
```diff
@@ -1312,18 +1309,30 @@ def scan_quality_check(label: str,
+                        pass
+                else:
+                    # Indicating neither increasing scan resolution nor freezing torisions is applied
+                    actions = {'inc_res': None}
```
`Used methods` are now changed to a list of `dict`s. The checking is a little bit complicated. We can simplify it to `actions = {'inc_res': None, 'freeze': 'all'}` in the future.

alongd replied:
Would you like to either make this change or add a `# Todo: ...` comment in the code?

alongd replied:
was this comment addressed?
(BTW, no need to write ""for developers" in docstings, all docstrings are for developers)

xiaoruiDong replied:
TODO comment is added and 'developer's are removed. Thanks

xiaoruiDong commented on arc/job/trsh.py:
```diff
@@ -1332,13 +1341,12 @@ def scan_quality_check(label: str,
-            actions = ['change conformer', pivots, deg_increment]
-            if actions in used_methods:
-                logger.error(f'Not troubleshooting a rotor with the same method: {actions}')
-                actions = list()
```
The check will be in `troubleshoot_scan_job`, which checks if the new conformer is used in the previous troubleshooting. I think this is more robust than check if this torsion has been rotated by XX degrees (equal `deg_increment` can still result in different conformers in different cases, if only the starting xyzs are not the same.

alongd commented:
@xiaoruiDong, I tried the branch out on some example, it seems to work very nicely, I'll uploade additional comments as I explore the output more.

Could you take a look at the failing tests?

Could you also add line breaks in the lengthy messages to make them readable, e.g.:
"Error: Rotor scan of s198 between pivots [4, 5] is inconsistent between two consecutive conformers. Between conformer # 6 and # 7, internal coordinates: D1, D2 changed significantly. Between conformer # 16 and # 17, internal coordinates: D7, D8, D9 changed significantly. ARC will attempt to troubleshoot this rotor scan."

alongd commented:
Also, we need to summarize the troubleshooting methods used per rotor so it is outputted for each species in the thermo library. Currently, we have the following description, for example:
"pivots: [1, 2], dihedral: [6, 1, 2, 3], rotor symmetry: 3, max scan energy: 7.34 kJ/mol".
The rotor description is set here: https://github.com/ReactionMechanismGenerator/ARC/blob/master/arc/statmech/arkane.py#L327

xiaoruiDong commented:
> @xiaoruiDong, I tried the branch out on some example, it seems to work very nicely, I'll upload additional comments as I explore the output more.
> 
> Could you take a look at the failing tests?
> 
> Could you also add line breaks in the lengthy messages to make them readable, e.g.:
> "Error: Rotor scan of s198 between pivots [4, 5] is inconsistent between two consecutive conformers. Between conformer # 6 and # 7, internal coordinates: D1, D2 changed significantly. Between conformer # 16 and # 17, internal coordinates: D7, D8, D9 changed significantly. ARC will attempt to troubleshoot this rotor scan."

Thank you for letting me aware of the test issues. It is an easy fix due to my recent change to the sequence of the input arguments.

For the error message, do you like the following pattern?
```
ERROR:arc:Rotor scan of  between pivots [2, 3] is inconsistent between two consecutive conformers.
Inconsistent consecutive conformers and problematicinternal coordinates:
conformer #  8/#  9        D21, D22, D23
conformer # 18/# 19        D21, D22, D23
conformer # 39/# 40        D21, D22, D23
ARC will attempt to troubleshoot this rotor scan.
```

alongd commented on arc/statmech/arkane.py:
```diff
@@ -323,10 +323,20 @@ def generate_arkane_species_file(self,
+                                scan_trsh = scan_trsh_method['scan_trsh']
+                            if scan_res > scan_trsh_method['scan_res']:
+                                scan_res = scan_trsh_method['scan_res']
+                    scan_trsh = f'Troubleshot with the following constraints and {scan_res} degrees ' \
```
I think this line should be further indented to be under the `if` condition above

xiaoruiDong replied:
Correct!!

alongd commented:
@xiaoruiDong, what's the status of this PR? I'd like to add TS rotors, but waiting for this to be merged first.

xiaoruiDong commented:
@alongd Sorry for the delay. Actually, I don't quite get what you suggest for the two non-addressed comments.. 

For the first one, do you want me to give an example? something like: already frozen [1,2,3,4], but [1,2,3,5] flip? 
For the second one, you mentioned adding  'TODO:', but there isn't any to-do, I think. 

I will clean everything up, as soon as the above addressed. Thanks!

alongd commented:
No problem.

1. Regarding the "todo" comment (line 1344 in trsh.py):
You wrote that we can simplify the checking ("`Used methods` are now changed to a list of `dict`s. The checking is a little bit complicated. We can simplify it to `actions = {'inc_res': None, 'freeze': 'all'}` in the future.")
I suggested we either simplify this in this PR as you suggested, otherwise if you think this should be done in the future, add a "Todo" comment in that line to document your intentions.

2. Regarding the example comment (line 626 in trsh.py):
Yes, if you could add to the function docsting something like the below, it would be great.
"For example: If ``scan = [1, 2, 3, 4]``, ``scan_quality_check()`` might find ``[1, 2, 3, 5]`` which is problematic (we shouldn't freeze ``[1, 2, 3, 5]`` in such case). By using this function, ARC can figure out which internal coordinates we need to actually freeze to troubleshoot the scan."

alongd commented:
Can you take a look at the Codacy report for the minor stuff?
https://app.codacy.com/manual/ReactionMechanismGenerator/ARC/pullRequest?bid=16680380&prid=5069484&utm_campaign=new_commit&utm_medium=Email&utm_source=Internal&

xiaoruiDong commented:
@alongd  check is finished. Thanks

alongd commented on arc/parser.py:
```diff
@@ -737,3 +774,267 @@ def process_conformers_file(conformers_path):
+            terms = line.split()
+            ic_dict['label'].append(terms[1])
+            ic_dict['type'].append(terms[1][0])  # 'R: bond, A: angle, D: dihedral
+            atom_inds = re.split(r'\(|\)|,', terms[2])[1:-1]
```
could `'\(|\)|,'` be replaced with `'[(),]'`?

xiaoruiDong replied:
Yes, your suggestion is clearer.

alongd commented on arc/scheduler.py:
```diff
@@ -2165,14 +2178,21 @@ def check_scan_job(self, label, job):
+        # causes the internal coordinates mess up in the middway of the scan. It can be resolved
+        # by conformer based scan troubleshooting method and its energies are readable.
+        if job.job_status[1]['status'] != 'done' \
+            and job.job_status[1]['error'] != 'Internal coordinate error':
```
this line should have an additional indentation

alongd commented on arc/scheduler.py:
```diff
@@ -2497,67 +2528,120 @@ def troubleshoot_negative_freq(self, label, job):
-    def troubleshoot_scan_job(self, job, methods=None):
+    def troubleshoot_scan_job(self,
+                              job: Job,
+                              methods: Optional[dict]=None,
```
minor: add spaces around `=`

alongd commented on arc/scheduler.py:
```diff
@@ -2497,67 +2528,120 @@ def troubleshoot_negative_freq(self, label, job):
+            # Check if the same conformer is used in previous troubleshooting
+            for used_trsh_method in used_trsh_methods:
+                if 'change conformer' in used_trsh_method \
+                    and compare_confs(new_xyz, used_trsh_method['change conformer']):
```
add indentation

alongd commented on arc/scheduler.py:
```diff
@@ -2497,67 +2528,120 @@ def troubleshoot_negative_freq(self, label, job):
+                if 'change conformer' in used_trsh_method \
+                    and compare_confs(new_xyz, used_trsh_method['change conformer']):
+                    # Find we have used this conformer for troubleshooting. Invalid the troubleshooting.
+                        logger.error(f'The change conformer method for {label} is invalid. ' \
```
remove one indent

alongd commented on arc/scheduler.py:
```diff
@@ -2497,67 +2528,120 @@ def troubleshoot_negative_freq(self, label, job):
-                self.run_job(label=label, xyz=job.xyz, level_of_theory=job.job_level_of_theory_dict, job_type='scan',
-                             scan=job.scan, pivots=job.pivots, scan_trsh=scan_trsh, scan_res=scan_res)
+                if scan_trsh or job.scan_res != scan_res \
+                    and {'scan_trsh': scan_trsh, 'scan_res': scan_res} not in used_trsh_methods:
```
indentation

xiaoruiDong replied:
It seems like my autopep8 using another rule, lol.

alongd commented on arc/scheduler.py:
```diff
@@ -2497,67 +2528,120 @@ def troubleshoot_negative_freq(self, label, job):
+                                 job_type='scan', scan=job.scan, pivots=job.pivots,
+                                 scan_trsh=scan_trsh, scan_res=scan_res)
+        return trsh_success, actual_actions
+
```
please remove extra line break

alongd commented on arc/scheduler.py:
```diff
@@ -2497,67 +2528,120 @@ def troubleshoot_negative_freq(self, label, job):
+        # A lower conformation was found
+        if 'change conformer' in methods:
+            # We will delete all of the jobs no matter we can successfully change to the conformer.
+            # If succeed, we have to cancel jobs to avoid conflictions
```
typo: "conflicts" instead of "conflictions"

xiaoruiDong commented on arc/common.py:
```diff
@@ -1092,3 +1094,89 @@ def estimate_orca_mem_cpu_requirement(num_heavy_atoms: int,
+    # First iteration without 180/-180 adjustment
+    change = (torsions[index_1] - torsions[index_2] - delta).abs() > threshold
+    # Apply 180/-180 adjustment to those shown significance
+    for label in change[change == True].index:
```
Actually, there are multiple steps. change is a table. The indexes are internal coordinates (IC) label, the value are `True` or `False`. 
`change[change == True]` will find all entries whose values are `True`
`change[change == True].index` is a list providing the IC labels of all these `True`s

xiaoruiDong commented on arc/common.py:
```diff
@@ -1092,3 +1094,89 @@ def estimate_orca_mem_cpu_requirement(num_heavy_atoms: int,
+                      change.
+    """
+    # First iteration without 180/-180 adjustment
+    change = (torsions[index_1] - torsions[index_2] - delta).abs() > threshold
```
> If index_1 and index_2 could be strings (according to the typing above), this would crash
Actually `torsions` is not a list but a DataFrame that can take 'str' or 'value' depending on what the column name. Other words, if we define the IC column names as numbers, then indexes should be numbers, and if we define them as `str`s, then indexes should be `str`

xiaoruiDong replied:
> I think the - delta should be outside the (...).abs()
or perhaps do: change = (torsions[index_1] - torsions[index_2]).abs() > delta + threshold

I think about it once, and I think it is better to include inside. Let me give you an example:
say [1,2,3,4] is being scanned at `step = 8` (which makes our `delta = 8`). I set default threshold as `20`. So, if delta is inside the parentheses, it means: 
```
abs(new_dihed - old_dihed - 8) > 20
# equivalent to
old_dihed - 12 < new_dihed < old_dihed + 28
```
otherwise, the boundary will be +- 28, which is less constrained. Considering the dihed is increasing, I think it is not a bad idea to constrain both sides differently.

alongd replied:
Thanks for the explanation, I agree.
Is it possible that in a case like `abs(new_dihed - old_dihed - 8) > 20`, `new_dihed` will be lower than `old_dihed`? For example, if `old_dihed = 355` and `new_dihed = 355 + 8 = -3`?
Alternatively, if the user sets the dihedral increment in ARC's settings to be negative (users might want to scan a specific molecule in "reverse"?), then a situation where `new_dihed = 50` and `old_dihed = 65` is problematic as well: it should be within the 20 degree tolerance, but this calculation will determine otherwise.

xiaoruiDong replied:
> Is it possible that in a case like abs(new_dihed - old_dihed - 8) > 20, new_dihed will be lower than old_dihed? For example, if old_dihed = 355 and new_dihed = 355 + 8 = -3?

For this case, the block below `if torsions.loc[label, index_1] * torsions.loc[label, index_2] < 0:` is aimed to solve this problem. If any has a sign flipping due to -180 <=> 180 transition, this function can handle them.

xiaoruiDong replied:
> Alternatively, if the user sets the dihedral increment in ARC's settings to be negative (users might want to scan a specific molecule in "reverse"?), then a situation where new_dihed = 50 and old_dihed = 65 is problematic as well: it should be within the 20 degree tolerance, but this calculation will determine otherwise.

Actually, `delta` is set as the same as `scan_res`, so if people do the reverse scan, they need to set `scan_res` negative, and `delta` will be negative automatically. Back to the example,
```
old_dihed = 65
step_size = -8
new_dihed_expect = 57
new_dihed = 50
```
50 and 57 have a difference of 7 within 20 degrees threshold, so this should not be problematic

In our code, if the following inequality is True, `abs(new_dihed - old_dihed - delta) > 20`, then we think the change is too huge and problematic. The inequality is equivalent to
```
new_dihed < old_dihed + 20 + delta
new_dihed > old_dihed - 20  + delta
# Plug-in our numbers
new_dihed < old_dihed + 20 + delta
new_dihed > old_dihed - 20  + delta
=>
new_dihed < old_dihed + 12
new_dihed > old_dihed - 28
=>
new_dihed < 65 + 20 - 8 = 77
new_dihed > 65 - 20 - 8 = 37
```
So it also works for negative resolution

xiaoruiDong commented on arc/common.py:
```diff
@@ -1092,3 +1094,89 @@ def estimate_orca_mem_cpu_requirement(num_heavy_atoms: int,
+               and abs(torsions.loc[label, index_1] + 360 - torsions.loc[label, index_2] - delta) < threshold:
+                change[label] = False
+            elif torsions.loc[label, index_2] < 0 \
+                 and abs(torsions.loc[label, index_1] - 360 - torsions.loc[label, index_2] - delta) < threshold:
```
okay

xiaoruiDong commented on arc/common.py:
```diff
@@ -18,8 +18,10 @@
 import time
 import warnings
 import yaml
+from typing import Optional, Union
```
I will

xiaoruiDong commented on arc/job/trsh.py:
```diff
@@ -501,49 +516,208 @@ def trsh_negative_freq(label: str,
         scan (list): The four atom indices representing the torsion to be troubleshooted.
-        species_scan_lists (list): Entries are lists of four atom indices each representing a torsion.
-        methods (list): The troubleshooting method/s to try. Accepted values: 'freeze' and/or 'inc_res'.
+        scan_lists (list): Entries are the four-atom scan lists (1-indexed) of all torsions
```
You are right

xiaoruiDong commented on arc/job/trsh.py:
```diff
@@ -501,49 +516,208 @@ def trsh_negative_freq(label: str,
+                #     just freeze the one first seen.
+                pruning = []
+                for torsion in problematic_ic:
+                    if torsion in pruning \
```
Thank you !

xiaoruiDong commented on arc/job/trsh.py:
```diff
@@ -1000,89 +1174,242 @@ def scan_quality_check(label: str,
+        bonds = scan_conformers[scan_conformers['type'] == 'R']
+        angles = scan_conformers[scan_conformers['type'] == 'A']
+        non_scan_rotor = scan_conformers[(scan_conformers['type'] == 'D') \
+                                         & (scan_conformers['scan'] == False)]
```
`scan_conformers` is similar to `torsions` mentioned above. Here, I am applying filters to scan_conformers. Let me briefly explain this line:

`non_scan_rotor = scan_conformers[(scan_conformers['type'] == 'D') & (scan_conformers['scan'] == False)]`

`scan_conformers` have many columns. This line filters based on two criteria: `'type'` and `'scan'`. Since `scan_conformers` has multiple rows, we cannot do `is` operation. For DataFrame, `scan_conformers['scan'] == False` is similar to the operation for np.NDArrays where each element will be compared. Besides, we should use `&` instead of using `and`, which I think is related to how DataFrame parses arguments.

alongd replied:
Thanks for the explanation!

xiaoruiDong commented on arc/job/trsh.py:
```diff
@@ -1000,89 +1174,242 @@ def scan_quality_check(label: str,
+        bonds = scan_conformers[scan_conformers['type'] == 'R']
+        angles = scan_conformers[scan_conformers['type'] == 'A']
+        non_scan_rotor = scan_conformers[(scan_conformers['type'] == 'D') \
+                                         & (scan_conformers['scan'] == False)]
```
See the explanation above. Thanks.

xiaoruiDong commented on arc/job/trsh.py:
```diff
@@ -1000,89 +1174,242 @@ def scan_quality_check(label: str,
+                        f'searching for a better conformation...'
+            logger.info(message)
+            # Find the dihedrals in degrees of the lowest conformer:
+            min_index = np.argmin(energies)
```
Let me describe what will be stored in the `actions`. 
1. I slightly modified when we save these `actions`, and now it is saved right after we find a troubleshooting strategy and ready to launch another job, (after `troubleshoot_scan_job()`). It used to be saved right after the `scan_quality_check()`. Although the thing generated from `scan_quality_check()` is also called `actions`, they are like raw actions, which will be tailored in the `troubleshoot_scan_job()`.

2.`actions` is a`dict` now, which will only have three kind of keys: `change conformer`, `scan_res` and `scan_trsh`. The value of `change conformer` is ARC's xyz dict. There is no `array`s. The `scan_res` will be either `float` or `int`, and `scan_trsh` is a `str`, they are just like what we had in the past.

xiaoruiDong commented on arc/job/trsh.py:
```diff
@@ -501,49 +516,208 @@ def trsh_negative_freq(label: str,
+                    pruning += [ic for ic in problematic_ic if is_same_pivot(torsion, ic)]
+        
+        # 2. Freeze all torsions other than the rotor to be scanned
+        if methods['freeze'] == 'all':
```
To make this function robust, I add an extra check in the `1.Freeze specific internal coordinates identified by scan_quality_check()` to see if we can successfully parse the scan arguments (scanned IC, frozen ICs, steps, etc). 
If succeed, `methods['freeze']` will be remained as a list. If not succeed, `methods['freeze']` will be changed to `'all'`, so that we can still do some sort of troubleshooting.

xiaoruiDong commented on arc/job/trsh.py:
```diff
@@ -1000,89 +1174,242 @@ def scan_quality_check(label: str,
+
+    # 1. Check based on intermediate conformers
+    if conformer_based_check:
+        bonds = scan_conformers[scan_conformers['type'] == 'R']
```
Actually, it is defined at line 1217. before the `try ... except ...`

alongd replied:
Sorry, I still didn't see where `scan_conformers` is defined. Line 1217 has:
```python
    conformer_based_check = False
```

xiaoruiDong replied:
Yep, the logic is if line 1220 `scan_conformers = parse_scan_conformers(log_file)` works okay, then it make `conformer_based_check = True`. Otherwise, `conformer_based_check = False`. For cases of `False` , line 1230 `if conformer_based_check:` will find out and won't go further to line 1231.

xiaoruiDong commented on arc/parser.py:
```diff
@@ -281,6 +284,40 @@ def parse_1d_scan_energies(path):
     return energies, angles
 
 
+def parse_1d_scan_coords(path):
```
If I understand correctly, `parse_1d_scan_coords()` returns xyz for optimized conformers, and 'parse_trajectory()` returns both optimized and non-optimized conformers. They can be merged, with an optional argument like 'optimized'. But I would suggest that we do it in another PR. I cherrypick `parse_1d_scan_coords()` from your commits and I just keep them intact. A change here may also increase your rebase and workload in TS rotor part.

xiaoruiDong commented on arc/parser.py:
```diff
@@ -737,3 +774,267 @@ def process_conformers_file(conformers_path):
+            # g09, g16
+            scan_blk = parse_str_blocks(file_path, 'The following ModRedundant input section has been read:',
+                                        'Isotopes and Nuclear Properties', regex=False)[0][1:-1]
+        except:
```
Yes, I agree it should be a more specific exception.

xiaoruiDong commented on arc/parser.py:
```diff
@@ -737,3 +774,267 @@ def process_conformers_file(conformers_path):
+        conformers.append(ics)
+    # Concatenate ICs of conformers to a single table and remove redundant ICs
+    scan_conformers = pd.concat([scan_ic_info] + conformers, axis=1)
+    red_ind = scan_conformers[scan_conformers.redundant == True].index
```
See explanation above

xiaoruiDong commented:
@alongd Alom the Rotor PR is updated. one of the checks is not passed because Travis dead. Is there a way to restart the check?

alongd commented:
I restarted the test, they all passed.
I added one comment re the definition of `scan_conformers`, please take a look

xiaoruiDong commented:
@alongd  I responded to the comments. Please find them out
